### PR TITLE
Catch authentication errors

### PR DESCRIPTION
**Pull request by kvchen**
Hides the tracebacks and prints a more user-friendly message when authentication fails under different scenarios (authorization code not received, access token is None). Addresses #152.

### CODE REVIEW

File: client/utils/auth.py [hunk 1]
```diff
@@ -61,12 +66,23 @@ def make_refresh_post(refresh_token):
...
+
+    access_token = storage['access_token']
+    expires_at = storage['expires_at']
+    refresh_token = storage['refresh_token']
+
+    return access_token, expires_at, refresh_token
+
 
 def update_storage(access_token, expires_in, refresh_token):
+    if not access_token and expires_in and refresh_token:
```
Comment (Sumukh) [inline client/utils/auth.py]: I think you want 

    if not (access_token and expires_in and refresh_token):

**Comment by Sumukh** [comment]
On cases where there is no code for us to exchange (meaning the permission request was denied) - we now send an error page in the browser
image

Terminal:
image

**Comment by Sumukh** [comment]
@kvchen - Anything else you want to add to this PR? 

I caught all exceptions when exchanging the token - which means that we shouldn't see any more tracebacks for students.

File: client/exceptions.py [hunk 2]
```diff
@@ -1,14 +1,18 @@
 """Client exceptions."""
 
+
```
Comment (Sumukh) [inline client/exceptions.py]: These blank lines are for PEP8 (E302)?

Comment (kvchen) [inline client/exceptions.py]: yeah - I didn't realize I'd committed those changes as well.

**Comment by kvchen** [comment]
Looks good - if we can narrow down the SSL issue later, we can open another issue.

File: client/utils/auth.py [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
 import time
-import webbrowser
+from urllib.parse import urlparse, parse_qs
 from urllib.request import urlopen
-import json
-from .html import auth_html, partial_course_html, partial_nocourse_html, red_css
+import webbrowser
+
+from client.exceptions import AuthenticationException
+from .html import auth_html, partial_course_html, partial_nocourse_html, \
```
Comment (albert12132) [inline client/utils/auth.py]: Avoid use relative imports: use `from client.utils.html import ...` instead.

File: client/utils/auth.py [hunk 4]
```diff
@@ -129,11 +154,26 @@ def authenticate(force=False):
...
-            access_token, refresh_token, expires_in = _make_code_post(code)
+
+            try:
+                code = qs['code'][0]
+                access_token, refresh_token, expires_in = _make_code_post(code)
+            except KeyError:
+                message = qs.get('error', 'Unknown')
+                log.warning("No auth code provided {}".format(message))
+                done = True
+                send_failure("{}".format(message))
```
Comment (albert12132) [inline client/utils/auth.py]: Just `send_failure(message)`. Or `str(message)` if message isn't necessarily a string.

File: client/utils/auth.py [hunk 5]
```diff
@@ -153,11 +193,32 @@ def log_message(self, format, *args):
...
+    return auth_html.format(
+        site=SERVER,
+        status=status,
+        courses=html,
+        byline=byline,
+        title=title,
+        head=head)
+
+def send_failure(message):
+    self.send_response(400)
```
Comment (albert12132) [inline client/utils/auth.py]: What is self in this context?

**Comment by Sumukh** [comment]
The changes to send_failure LGTM. 

@albert12132 : Any other comments? If not, I'll deploy this as `v1.5.2`

**Comment by albert12132** [comment]
Sounds good.

**Pull request merged**